protocol-specs/generators/dpos_random_seed_generation/index.js

```diff
@@ -0,0 +1,433 @@
+};
+
+const generateSeedReveals = ({ delegateList, numberOfBlocks }) => {
+	const seeds = {};
+
+	for (const delegate of delegateList) {
+		const seedsForDelegate = generateSeedOnion(
+			delegate.publicKey,
+			numberOfBlocks,
+		);
+		const counter = 0;
+
+		seeds[delegate.publicKey] = {
+			counter,
+			seeds: seedsForDelegate,
+		};
+	}
+
+	return seeds;
+};
+
+const generateBlocks = ({ startHeight, numberOfBlocks, delegateList }) => {
+	const seedReveals = generateSeedReveals({ delegateList, numberOfBlocks });
+	const numberOfDelegates = delegateList.length;
+
+	return new Array(numberOfBlocks).fill(0).map((_v, index) => {
+		const height = startHeight + index;
+		const { publicKey } = delegateList[index % numberOfDelegates];
+		const seedReveal =
+			seedReveals[publicKey].seeds[seedReveals[publicKey].counter];
+
+		seedReveals[publicKey].counter += 1;
+
+		return {
+			generatorPublicKey: delegateList[index % numberOfDelegates].publicKey,
+			height,
+			seedReveal,
+		};
+	});
+};
+
+const round = (height, n) => Math.floor((height - 1) / n) + 1;
+
+const roundFlags = (r, n) => {
+	const firstBlock = r * n - n + 1;
+	const lastBlock = r * n;
+	const middleBlock = Math.floor((firstBlock + lastBlock) / 2);
+
+	return {
+		round: r,
+		firstBlock,
+		lastBlock,
+		middleBlock,
+	};
+};
+
+const generateRandomSeed = (blocks, blocksPerRound) => {
+	// Last block height
+	const l = blocks[blocks.length - 1].height;
+
+	// Number of blocks in a round
+	const n = blocksPerRound;
+
+	// Current round
+	const r = roundFlags(round(l, n), n);
+
+	// First block of current round
+	const { firstBlock: h } = r;
+
+	// Middle range of a round
+	const m = Math.floor(n / 2);
+
+	if (l < r.middleBlock) {
+		throw new Error(
+			`Random seed can't be calculated earlier in a round. Wait till you pass middle of round. Current height ${l}`,
+		);
+	}
+
+	if (l >= r.middleBlock && r.round === 1) {
+		const randomSeed1 = H(m + 1);
+		const randomSeed2 = H(0);
+
+		return { randomSeed1, randomSeed2 };
+	}
+
+	const { firstBlock: firstBlockBlockOfLastBlockRound } = roundFlags(
+		r.round - 1,
+		n,
+	);
+
+	const filteredBlocks = blocks.reduce((acc, b) => {
+		if (b.height >= firstBlockBlockOfLastBlockRound) {
+			acc[b.height] = b;
+		}
+		return acc;
+	}, {});
+
+	const seedRevealMap = Object.values(filteredBlocks).reduce((acc, b) => {
+		acc[b.height] = b.seedReveal;
+		return acc;
+	}, {});
+
+	const forgedBlocksMap = Object.values(filteredBlocks).reduce((acc, b) => {
+		if (!acc[b.generatorPublicKey]) {
+			acc[b.generatorPublicKey] = [];
+		}
+		acc[b.generatorPublicKey].push(b.height);
+		return acc;
+	}, {});
+
+	const seedRevealsForRandomSeed1 = [];
+
+	// From middle of current round to middle of last round
+	for (let i = h + m; i >= h - m; i -= 1) {
+		const { generatorPublicKey, seedReveal } = filteredBlocks[i];
+		const [previousForgedHeight] = forgedBlocksMap[generatorPublicKey];
+
+		// If forged only 1 block in current and last round
```

Comment:
this comment may be misleading.  It should be something like `If forged only 1 block in round at height i or previous round`. We have to do the seed reveal check respect to the considered block.